### ISSUE TITLE
using open() without close()

### ISSUE THREAD

**Issue by airallergy** [pyenergyplus/eppy3000#60]
I saw a few cases that use `open()` without a `with ... as` statement or a following `close()`, I suppose this could be not safe, and could drag down the performance since those schema files are quite large.

**Comment by santoshphilip** [comment]
At this point eppy3000 opens three kinds of  files:
1. epJSON file
2. epSchema file
3. idf files (to convert to epJSON)

We should check ` close` or `with` in all three cases.

The functions should be able  to be able to open 
- files from the disk 
- files in memory (StringIO)

Right now it is a bit of a mix. Some functions open files from disk and some functions can open files in memory. (some parts of the API are broken because of this)
This should be cleaned up so that it can identify what kind of file it is and can open it appropriately. `Eppy` does this reasonably well.

**Comment by airallergy** [comment]
In terms of files from the disk, when involving `pathlib`, an elegant way to do is

```python 
from pathlib import Path

file = Path(...) / ... / ...
with file.open(...) as fh:
    ...
```

Let me try to amend them in a few days.

**Comment by airallergy** [comment]
One question, should you create another develop branch, or should I pull request directly to the master?

**Comment by santoshphilip** [comment]
Do it directly to master.
I agree on the elegance of pathlib - it is a pleasure to use

**Comment by airallergy** [comment]
How do you prefer if I have some general questions to ask about the codes that are essentially neither new issues nor related to an existing one?

**Comment by airallergy** [comment]
btw, do you mind having all codes run through black (if thats still what you intend to use for formatting)? otherwise I found that git diffs would be very ugly...

**Comment by santoshphilip** [comment]
regarding discussion issues. 
- An issue can be just a discussion - does not have to be a problem
- If it is really off topic, email me at santosh_philip@yahoo.com
    - I'll delete this comment once you have emailed me
    - So my email is not sitting around

**Comment by santoshphilip** [comment]
I'll run black.
I have been coding fast and hard at this early stage. So things have been falling on the wayside.

**Comment by airallergy** [comment]
wondering what editor you are using, there might be a function to auto format on save/type/other conditions?

**Comment by airallergy** [comment]
A further issue is about the support of `io.StringIO` and `io.TextIOWrapper`. I know that they are very convenient in the unit test, but are there use cases other than that? should we expect that users would utilise them?

The issues are that they are intended to use along with `close()`, making it troublesome to be mixed with `with ... as`, which is a preferred way to deal with io these days I suppose. Two-ish options I have in mind:

1. keep supporting them, 

    a. separate them from the reading disk files, they have a `.close()` in the end whilst disk files use `with ... as`

    b. no use of `with ... as` at all

2. stop supporting them, restrict users to disk files only, and make some workarounds in unit tests, such as using a real epjson file, or construct a fake file from string using `pytest`

basically, the first one complicates the source codes and the second the unit test codes. The key point I suppose is whether this is a wanted function to users. What do you think?

I am also wondering what the initial intention is to include them in `eppy`? There may be some considerations that I have not realised myself.

**Comment by santoshphilip** [comment]
I understand the point you are bringing up.
I had put a fair amount of thought into this matter before, but I don't remember much of it :-(

My best recollection:

- there is no guarantee that the data will come from a disk.
    - it can come from a socket or stream or pipe
    - I think zeppy <https://github.com/pyenergyplus/zeppy> works with files coming from a socket or a pipe. Zeppy uses eppy
- Ideally we should work with a file-like object <https://docs.python.org/3/glossary.html#term-file-like-object>
- Many of the python modules work with file-like objects. See the following modules:
    - json <https://pymotw.com/3/json/index.html#working-with-streams-and-files>
    - csv 
    - pickle
    - xml.etree.ElementTree
    
We are really looking at API design here. Not the coding or the testing. 

- We could write the API to work with file-like objects
- Then have a wrapper functions that does files from disk
    - Most of our use-case is files from disk
    - So we should make it easy to do that
    
`eppy`'s API is not optimal. We don't need to stick to that. It happens to be good starting point, so we don't have to reinvent anything. 

Anyways, research it a bit and let me know where it takes you.

**Comment by santoshphilip** [comment]
I prefer using black at just before the pull request.
I tend to use fairly simple editors (textmate and vi on mac/unix and Notepad++ or atom on PC). I like to have what I typed stay where I typed it. If things move every time I save, it messes with the pattern recognition part of my brain.

    On Saturday, August 8, 2020, 10:37:11 AM PDT, Cheng Cui <notifications@github.com> wrote:  
 
 


wondering what editor you are using, there might be a function to auto format on save/type/other conditions?

**Comment by airallergy** [comment]
It is a very interesting subject to look at, I have been using `json` a lot but never noticed the relationship between `load` and `loads`.

I think you are right about working with file like objects with a wrapper to deal with disk files, which I believe can make the structure clearer and less complicated. What is on my mind right now:

```python
# in eppy3000.modelmaker
class EPJ:
    def __init__(self, epj_fh, epschema_fh=None):
        self.epj = parse(epj_fh)
        self.epschema = parse(epschema) if epschema_fh else None
        self.epj_filename = None
        self.epschema_filename = None
```

```python
# in eppy3000.__init__
def epj(epj, epschema):
    if epj is os.PathLike: # pseudocode
        epj_fh = open(epj, "rt")
        epschema_fh = open(epschema, "rt") if epschema else None
        epj = EPJ(epj_fh, epschema_fh)
        epj_fh.close()
        if epschema: epschema_fh.close()
        epj.epj_filename = epj
        epj.epschema_filename = epschema
        return epj
    else:
        return EPJ(epj_fh, epschema_fh)
    
```

1. these codes are typed directly in the comment box, so just an illustration of ideas

2. it might not be good to process epw along with epj and epschema, because it is essentially not part of epj, and only supposed to be used when running ep if I am not mistaken, and clearly the expected input type is different (epw should accepts only a file path or a link I suppose, instead of a file-like object

3. this structure is something like `np.array` to `np.ndarray`

4. if the inputs are file-like objects instead of file paths, it should be users' responsibility to close them properly, as how `json`, `csv`, `pickle` do it, something like
    ```python
    with open("path/to/epj/file", "rt") as fh:
        epj = eppy3000.epj(fh)
    ```
    or
    ```python
    fh = io.StringIO("epj string")
    epj = eppy3000.epj(fh)
    fh.close()
    ```
    should be expected

5. I think we should separate the variable for epj itself from that for epj file path, for the sake of the implementation of functions like `save`, `run`, etc.

6. is there any benefit here to convert file paths to `pathlib.Path` objects? I failed to see one. pathlib is quite useful to construct platform dependent paths, say automatically find the installed ep executable, but here the paths seem to be static and provided by users

btw, when you said `"""same as modelmaker in eppy"""` in `eppy3000.modelmaker`, do you just mean `modeleditor`?

**Comment by santoshphilip** [comment]
Response to 1. 
I like your code snippets and the thinking behind it. 
I am inclined to hand off that part of the code to you.

I have created a branch `i60_fileobject` from the master.
Can you make the pull requests to that branch for this issue
You are welcome to break things there as you figure out the right way to do it.

Response to 2.
Agree on epw. Let us leave it out for the moment until we write the E+   runner

Response to 3. 
Have not used numpy much. Can you share some links.

Response to 4. 
agree

Response to 5. 
agree

Response to 6. 
Not seeing a benefit off the top of my head.

I misnamed modeleditor when I started work. Now it is stuck. We can roll back to modelmaker if needed.

**Comment by airallergy** [comment]
Sure thing, this might take some time.

As to numpy, it is hard to give you links to the codes, since most of the core codes are written in c as you may know. But the idea is that `numpy.array` itself isn't a class, it is just a helper function to construct a `numpy.ndarray` object, e.g.

```python
>>>import numpy as np
>>>a = np.array([1, 2, 3])
>>>type(a)
<class 'numpy.ndarray'>
```

which resembles what we are trying to do here.

**Comment by santoshphilip** [comment]
Don't do anything more on numpy. That was just a comment.
I'll figure it out on my own.

**Comment by airallergy** [comment]
@santoshphilip I'm inclined to change the variable `epj` to `ep` where relevant. The motivation is to differentiate it from the class `EPJ` (which represents the epJSON files consisting of the ep file and the ep schema file), and also to keep consistent with the variable `epschema` (since the dominating file format in `eppy3000` is epJSON, which seems no need to be explicitly stated). When there is a need to convert to idf or other formats, we can then add a postfix to distinguish, like `ep_idf`.

What do you think?

**Comment by santoshphilip** [comment]
That sounds reasonable. 

I am doing some work that may break when you do that.
Can you hold off until I sync. my updates

**Comment by airallergy** [comment]
Yeah sure, go ahead

**Comment by airallergy** [comment]
Also, if you have time, would you mind to briefly illustrate/describle the current workflow of the eppy3000 i/o (from read to save)? so that I can make sure I fully understand the logic and can better discuss it with you.

Currently the read and write function have different processes, and the write function seems like a workaround (with a lot of back forth conversions, todict, tomunch, tojson). I am trying to find if there is a way to simplify this and unify them, and would like to discuss with you a bit.

**Comment by airallergy** [comment]
Hi @santoshphilip, sorry for being idle for a long time, I was massively occupied by my own work, resulting in very limited spare time unfortunately.

I have been intensively using `eppy` for the past few years, which leads me to think that what was initially proposed here (i.e. following the `json` library style) does not make sense to me any more, as it may be too complex/verbose for `eppy` users. The `openidf` function in `eppy` is great, I wasn't aware of it back then, but have been relying on it since noticing it in most situations. Apparently I have no clue what updates you have made, if there ever is a need to think of an alternative way of reading idfs/epjsons (compared to `eppy`), my two cents is the `pandas` approach (i.e. class methods in the `IDF`/`EPJ` class, e.g. `from_epjson`) may be a better way.

I am happy if you'd like to close this issue and delete the branch.

**Comment by santoshphilip** [comment]
I would be interested in your thinking on this.
Since you probably use eppy more than I do at this point - you may have a better insight on this.

I have come to certain conclusions on the nature of JSON format.
- json format claims to he human readable. It is not very readable for normal humans.
- I am looking into YAML and TOML format and including them into eppy3000
    - prefer TOML since I like the name :-)
- Plan to  continue using IDF as a text file format
- maybe, use JSON only internally without exposing it to the interface API ?
    - makes sense to use JSON internally since that is the file structure that E+ uses

My observations:

- JSON is hard to edit in a text editor
- IDF, YAML and TOML are much easier to view, understand and edit in text

On Pandas:

- I would like to see your thinking on the pandas approach.
- I have used pandas only sporadically. So my understanding of pandas is not very deep

Maybe we both make a branch/fork and write some demo implementations of what we are thinking about ? It is easier to talk about things when you can see and play with it.

**Comment by santoshphilip** [comment]
Let us continue this conversation in this thread and close it later.

**Comment by airallergy** [comment]
I am not sure if I understand the intention of introducing another file format, we have JSON/epJSON on the energyplus side, and JSON/dict(?) inside eppy3000, where is a new format used?

That being said, I totally agree with your thoughts about JSON's poor readability! I don't know TOML before, but I happen to have had some thinking about whether to use JSON or YAML for a project of mine, and ended up favouring YAML for the same reasoning (I haven't done anything yet).

Regarding pandas, it is nothing about what is behind pandas, I basically meant the `from_`/`to_` + IO type pattern, some IO types may be dependant on one another. I personally found this pattern is very straightforward, but maybe I'm biased. I have not thought it through, what jumps to my mind is (based on my very vague memory about eppy3000 and our previous conversion)

```python
class EPJ:
    ...
    @classmethod
    def from_StringIO(cls, epj_fileobj):
        epj = json.load(epj_fileobj)
        ...
        return cls(...)

    @classmethod
    def from_epJSON(cls, epj_file):
        with open(epj, "rt") as fp:
            return from_StringIO(cls, fp)
```

I have to mention that my spare time is still limited due to the ongoing work, but I'd be happy to find time to write some demo implementations if this conversation goes to somewhere :)
